### PR TITLE
Hostagent code registers  for Snat, add update and delete events from…

### PR DESCRIPTION
**Pull request by jayaramsatya**
… Snat Operator controller based on notifications

it creates/deletes Snat json file in this directory((/usr/local/var/lib/opflex-agent-ovs/snats/),
Opflex agent  watches for the changes for this snat file to create the forward,reverse and bounce flows,
Snat file has 2 sections to represent the Local and remote data following is the file structure

{
    "uuid": "c47d3a21-b112-4516-ba2f-5e159b6c9e77",
    "snat-ip": "10.3.56.2",
    "interface-name": "of-cde81812ffc0",
    "interface-mac": "fa:16:3e:ba:97:b7",
    "interface-vlan": 1024,
    "local": true,
    "dest": "100.100.100.0",
    "dest-prefix": 24,
    "port-range":  [
        {
               "start": 32768,
                "end": 32800
        }
    ]
    "remote": {
        "port-map": [
             {
                               "mac": "fa:16:3e:ba:97:b7",`
                "port-range": [
                 {
                     "start": 40000,
                     "end": 40024
                }
            ]
              },
             {
                "mac": "fa:16:3e:ba:97:b8",`
                "port-range": [
                 {
                     "start": 40050,
                     "end": 40074
                }
            ]
              }
        ]
    }
}
Host agent watches for local and global info events. local info is maintained per node and global info is maintained across the nodes, local info event will be triggered whenever the POD ADD/DELETE happens, Hostagent maintains the state of this local information to update the EP file appropriately. And Global Info trigger comes whenever there is first pod comes in that node with snatIP association, and Hostagent maintains Global Info state to update the snat file

### CODE REVIEW

**Comment by mchalla** [comment]
Is it possible for you to skip the generated code and generate a diff ?

File: pkg/hostagent/opflex.go [hunk 1]
```diff
@@ -110,9 +110,13 @@ func (agent *HostAgent) discoverHostConfig() (conf *HostAgentNodeConfig) {
...
 			conf.VxlanAnycastIp = anycast.String()
 			conf.OpflexPeerIp = peerIp.String()
-			return
 		}
 	}
+	intf, err := net.InterfaceByName(conf.UplinkIface)
+	if err == nil {
+		conf.UplinkMacAdress = intf.HardwareAddr.String()
+		agent.log.Info("Hardware address is ", intf.HardwareAddr)
+	}
```
Comment (snaiksat) [inline pkg/hostagent/opflex.go]: Should we proceed further if we don't know the mac address?

File: pkg/hostagent/pods.go [hunk 2]
```diff
@@ -191,7 +192,21 @@ func (agent *HostAgent) syncEps() bool {
...
 					continue
 				}
-
+				agent.indexMutex.Lock()
+				agent.log.Debug("snat local info", agent.opflexSnatLocalInfos)
+				for k, v := range agent.opflexSnatLocalInfos {
+					if k == poduuid {
+						if agent.opflexSnatLocalInfos[poduuid].MarkDelete == true {
+							delete(agent.opflexSnatLocalInfos, poduuid)
+						   	ep.SnatIp = ""
```
Comment (snaiksat) [inline pkg/hostagent/pods.go]: Perhaps good to debug log here also?

File: pkg/hostagent/pods.go [hunk 3]
```diff
@@ -191,7 +192,21 @@ func (agent *HostAgent) syncEps() bool {
 				if ep.Uuid != epidstr {
 					continue
 				}
-
+				agent.indexMutex.Lock()
```
Comment (snaiksat) [inline pkg/hostagent/pods.go]: Should this processing happen in code block L146-L152?

File: pkg/hostagent/snats_test.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
+// Copyright 2017 Cisco Systems, Inc.
+//
```
Comment (snaiksat) [inline pkg/hostagent/snats_test.go]: Please update date

File: pkg/hostagent/snats.go [hunk 5]
```diff
@@ -0,0 +1,476 @@
+// Copyright 2016 Cisco Systems, Inc.
```
Comment (snaiksat) [inline pkg/hostagent/snats.go]: Please update date

File: pkg/hostagent/snats.go [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+		for _, ginfo := range v {
+			if nodename != agent.config.NodeName {
+				var remote OpflexSnatIpRemoteInfo
+				remote.MacAddress = ginfo.MacAddress
+				remote.PortRange = ginfo.PortRange
+				remoteinfo[ginfo.SnatIp] = append(remoteinfo[ginfo.SnatIp], remote)
+			}
+		}
+	}
+	agent.log.Debug("Remte Info: ", remoteinfo)
```
Comment (snaiksat) [inline pkg/hostagent/snats.go]: Remote

File: pkg/hostagent/snats.go [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+			syncSnat = true
+		}
+	}
+	agent.log.Debug("Snat Gobal Obj Map: ", agent.opflexSnatGlobalInfos)
+	if syncSnat {
+		agent.scheduleSyncSnats()
+	}
+}
+func (agent *HostAgent) snatLocalInfoDelete(obj interface{}) {
+	agent.log.Debug("Snat Delete Obj")
```
Comment (snaiksat) [inline pkg/hostagent/snats.go]: make this log message more specific (local)

File: pkg/hostagent/snats.go [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+			}
+		}
+	}
+	if syncep {
+		agent.scheduleSyncEps()
+	}
+}
+
+func (agent *HostAgent) snatGlobalInfoDelete(obj interface{}) {
+	agent.log.Debug("Snat Delete Obj")
```
Comment (snaiksat) [inline pkg/hostagent/snats.go]: make this log message more specific (global)

File: pkg/hostagent/snats.go [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+	// This case is true when scope moves from namespace to deployment
+	if len(localInfo) < len(agent.opflexSnatLocalInfos) {
+		for poduid, v := range agent.opflexSnatLocalInfos {
+			if _, ok := localInfo[poduid]; !ok {
+				// if pod present mark it snat ip deleted for local info
+				if _, ok := agent.opflexEps[poduid]; ok {
+					v.MarkDelete = true
+					syncep = true
+				} else {
+					delete(agent.opflexSnatLocalInfos, poduid)
```
Comment (snaiksat) [inline pkg/hostagent/snats.go]: There are changes to the cache happening here and below, you don't need to acquire lock?

File: pkg/hostagent/snats.go [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+	if logger == nil {
+		logger = agent.log.WithFields(logrus.Fields{})
+	}
+	logger.Debug("Snat Global info Changed...")
+	globalInfo := snat.Spec.GlobalInfos
+	// This case is possible when all the pods will be deleted from that node
+	if len(globalInfo) < len(agent.opflexSnatGlobalInfos) {
+		for nodename, _ := range agent.opflexSnatGlobalInfos {
+			if _, ok := globalInfo[nodename]; !ok {
+				delete(agent.opflexSnatGlobalInfos, nodename)
```
Comment (snaiksat) [inline pkg/hostagent/snats.go]: There are changes to the cache happening here, you don't need to acquire lock?

File: pkg/hostagent/agent.go [hunk 11]
```diff
@@ -44,30 +44,32 @@ type HostAgent struct {
...
+	podInformer        cache.SharedIndexInformer
+	endpointsInformer  cache.SharedIndexInformer
+	serviceInformer    cache.SharedIndexInformer
+	nodeInformer       cache.SharedIndexInformer
+	nsInformer         cache.SharedIndexInformer
+	netPolInformer     cache.SharedIndexInformer
+	depInformer        cache.SharedIndexInformer
+	rcInformer         cache.SharedIndexInformer
+	snatLocalInformer  cache.SharedIndexInformer
+	snatGlobalInformer cache.SharedIndexInformer
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/agent.go]: I thought all the informers can be in one place

File: pkg/hostagent/agent.go [hunk 12]
```diff
@@ -191,7 +201,8 @@ func (agent *HostAgent) Run(stopCh <-chan struct{}) {
 	}
 
 	if agent.config.OpFlexEndpointDir == "" ||
-		agent.config.OpFlexServiceDir == "" {
+		agent.config.OpFlexServiceDir == "" ||
+		agent.config.OpFlexSnatDir == "" {
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/agent.go]: Changed

File: pkg/hostagent/config.go [hunk 13]
```diff
@@ -61,6 +61,9 @@ type HostAgentNodeConfig struct {
 
 	// Anycast IP used for OpFlex communication
 	OpflexPeerIp string `json:"opflex-peer-ip,omitempty"`
+        
+	// Uplink Mac adress
+        UplinkMacAdress string `json:"uplink mac_adress,omitempty"`
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/config.go]: This is capture the Node Mac address

File: pkg/hostagent/config.go [hunk 14]
```diff
@@ -104,6 +107,9 @@ type HostAgentConfig struct {
 
 	// Directory for writing OpFlex service metadata
 	OpFlexServiceDir string `json:"opflex-service-dir,omitempty"`
+	
+	// Directory for writing OpFlex service metadata
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/config.go]: Changed

File: pkg/hostagent/config.go [hunk 15]
```diff
@@ -181,6 +187,7 @@ func (config *HostAgentConfig) InitFlags() {
 	flag.StringVar(&config.OpFlexConfigPath, "opflex-config-path", "/usr/local/etc/opflex-agent-ovs/base-conf.d", "Directory for writing Opflex configuration")
 	flag.StringVar(&config.OpFlexEndpointDir, "opflex-endpoint-dir", "/usr/local/var/lib/opflex-agent-ovs/endpoints/", "Directory for writing OpFlex endpoint metadata")
 	flag.StringVar(&config.OpFlexServiceDir, "opflex-service-dir", "/usr/local/var/lib/opflex-agent-ovs/services/", "Directory for writing OpFlex anycast service metadata")
+	flag.StringVar(&config.OpFlexSnatDir, "opflex-snat-dir", "/usr/local/var/lib/opflex-agent-ovs/snats/", "Directory for writing OpFlex anycast snat metadata")
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/config.go]: Changed

File: pkg/hostagent/environment.go [hunk 16]
```diff
@@ -85,7 +91,22 @@ func NewK8sEnvironment(config *HostAgentConfig, log *logrus.Logger) (*K8sEnviron
...
+		return nil, err
+	}
+	snatGlobalClient, err := snatglobalclset.NewForConfig(restconfig)
+	if err != nil {
+		return nil, err
+	}
+	nodeInfo, err := nodeinfoclientset.NewForConfig(restconfig)
+	log.Debug("Initializing kubernetes client", nodeInfo)
+	if err != nil {
+		log.Debug("Failed to intialize node info client")
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/environment.go]: Changed

File: pkg/hostagent/environment.go [hunk 17]
```diff
@@ -113,24 +136,28 @@ func (env *K8sEnvironment) PrepareRun(stopCh <-chan struct{}) (bool, error) {
 
 	env.agent.log.Debug("Starting node informer")
 	go env.agent.nodeInformer.Run(stopCh)
-
 	env.agent.log.Info("Waiting for node cache sync")
 	cache.WaitForCacheSync(stopCh, env.agent.nodeInformer.HasSynced)
 	env.agent.log.Info("Node cache sync successful")
 
 	env.agent.log.Debug("Starting remaining informers")
+	env.agent.log.Debug("exporting NodeInfo: ", env.agent.config.NodeName)
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/environment.go]: Changed, still Node info is not available here so only printing  for node name we are collecting.in subsequent log mac address will be displayed

File: pkg/hostagent/integ_test.go [hunk 18]
```diff
@@ -129,11 +129,16 @@ func SetupInteg(t *testing.T, c *HostAgentConfig) *integ {
 	if err != nil {
 		panic(err)
 	}
+	snatdir, err := ioutil.TempDir("", "it_snat_")
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/integ_test.go]: I have just followed it as above. as this just to create temporary directory creation for testcase

File: pkg/hostagent/nodeinfo.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+func (agent *HostAgent) InformNodeInfo(nodeInfoClient *nodeinfoclientset.Clientset,
+	kubeClient *kubernetes.Clientset) {
+	if nodeInfoClient == nil || kubeClient == nil {
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/nodeinfo.go]: Changed

File: pkg/hostagent/nodeinfo.go [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+	label := make(map[string]string, 0)
+	label["name"] = "aci-containers-host"
+	options := metav1.ListOptions{
+		LabelSelector: labels.Set(label).String(),
+		FieldSelector: fields.Set{"spec.nodeName": agent.config.NodeName}.String(),
+	}
+	existingPods, err1 := kubeClient.Core().Pods("").List(options)
+
+	if err1 != nil {
+		agent.log.Debug("failed to list existing pods in the podSet", err1)
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/nodeinfo.go]: changed based on comment

File: pkg/hostagent/nodeinfo.go [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+			Namespace: namespace,
+		},
+		Spec: nodeinfov1.NodeinfoSpec{
+			Nodename:   agent.config.NodeName,
+			Macaddress: agent.config.UplinkMacAdress,
+		},
+	}
+	result, err := nodeInfoClient.AciV1().Nodeinfos(namespace).Create(nodeInfoInstance)
+	if err == nil {
+		agent.log.Debug("NodeInfo CR is created", result)
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/nodeinfo.go]: changed based on comment

File: pkg/hostagent/nodeinfo.go [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+	} else if apierrors.IsAlreadyExists(err) {
+		agent.log.Debug("Already exists: %#\n", result)
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/nodeinfo.go]: changed based on comment

File: pkg/hostagent/nodeinfo.go [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+	} else {
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply pkg/hostagent/nodeinfo.go]: Added log

File: pkg/hostagent/opflex.go (hunk 1 above)
Comment (jayaramsatya) [reply pkg/hostagent/opflex.go]: return is added

File: pkg/hostagent/pods.go (hunk 3 above)
Comment (jayaramsatya) [reply pkg/hostagent/pods.go]: yes

File: pkg/hostagent/pods.go (hunk 2 above)
Comment (jayaramsatya) [reply pkg/hostagent/pods.go]: add the log

File: pkg/hostagent/snats.go (hunk 9 above)
Comment (jayaramsatya) [reply pkg/hostagent/snats.go]: calling function acquired the lock after that it is deferred the lock until the function execution completes

File: pkg/hostagent/snats.go (hunk 5 above)
Comment (jayaramsatya) [reply pkg/hostagent/snats.go]: Updated

File: pkg/hostagent/snats_test.go (hunk 4 above)
Comment (jayaramsatya) [reply pkg/hostagent/snats_test.go]: Updated

File: docker/launch-hostagent.sh [hunk 24]
```diff
@@ -34,6 +34,7 @@ if [ -w ${PREFIX} ]; then
     # Setup folders to hold metadata
     mkdir -p ${VARDIR}/lib/opflex-agent-ovs/endpoints
     mkdir -p ${VARDIR}/lib/opflex-agent-ovs/services
+    mkdir -p ${VARDIR}/lib/opflex-agent-ovs/snats
```
[the remark this answers is not in the document]
Comment (jayaramsatya) [reply docker/launch-hostagent.sh]: Just followed the convention of endpoint and service files. because we will create multiple files in snat  per IP

**Pull request merged**